llvm/lib/CodeGen/BranchFolding.cpp

```diff
@@ -85,6 +85,11 @@ TailMergeSize("tail-merge-size",
               cl::desc("Min number of instructions to consider tail merging"),
               cl::init(3), cl::Hidden);
 
+static cl::opt<bool> TailMergeOnlyBBsWithoutSucc(
+    "tail-merge-only-bbs-without-succ",
+    cl::desc("Tail merge only basic blocks without successors"),
+    cl::init(false), cl::Hidden);
```

Comment:
What is the point of the flag? Generally cl::opt should only be for test/debug, but I don't see how that is useful here.  If a target wants this behavior, it should be a target hook

Reply:
This flag should be used like other flags (e.g. `-tail-merge-size`, `-tail-dup-size`) for specific applications to get maximum possible performance, and in this case target hook won't help. Since `-enable-tail-merge=false` disables tail merge for all BBs, I introduced this option to just enable tail merge for exiting BBs (which are not frequently executed), so we can get code size reduction.
Does this make sense to you?

Reply:
Can you use branch probabilities to decide which are not frequently executed?

> Does this make sense to you?

I mean in terms of adding a new debug flag it's no worse than anything else already here, but none of these are really intended for end users

Reply:
> Can you use branch probabilities to decide which are not frequently executed?

I think it is feasible, but taking into account that here we can just use early return if option is passed, I would say it is not worth it to add complexity to this.

Reply:
Options are complexity, not discoverable, and cl::opts are not really for user facing options

Reply:
Can you drive this off of minsize attribute?

Reply:
From what you suggested:
> If a target wants this behavior, it should be a target hook
Are you OK to implement this approach, so I can override it for a downstream target I'm working on?